cmd/snap-confine/udev-support.c

```diff
@@ -197,6 +204,37 @@ void setup_devices_cgroup(const char *security_tag, struct snappy_udev *udev_s)
 	for (int i = 0; static_devices[i] != NULL; i++)
 		run_snappy_app_dev_add(udev_s, static_devices[i]);
 
+	// nvidia modules are proprietary and therefore aren't in sysfs and
+	// can't be udev tagged. For now, just add existing nvidia devices to
+	// the cgroup unconditionally (AppArmor will still mediate the access).
+	// We'll want to rethink this if snapd needs to mediate access to other
+	// proprietary devices.
+	struct stat sbuf;
+	char nvpath[16] = { 0 };	// /dev/nvidiaXXX, ctl and -uvm
+	for (unsigned minor = 0; minor < 255; minor++) {
+		// https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/devices.txt
+		// /dev/nvidia0 through /dev/nvidia254
+		sc_must_snprintf(nvpath, sizeof(nvpath), "/dev/nvidia%u",
+				 minor);
+
+		// Stop trying to find devices after one is not found. In this
+		// manner, we'll add /dev/nvidia0 and /dev/nvidia1 but stop
+		// trying to find nvidia3 - nvidia254 if nvidia2 is not found.
+		if (stat(nvpath, &sbuf) != 0) {
+			break;
+		}
+		_run_snappy_app_dev_add_majmin(udev_s, nvpath, 195, minor);
+	}
+
+	sc_must_snprintf(nvpath, sizeof(nvpath), "/dev/nvidia%s", "ctl");
+	if (stat(nvpath, &sbuf) == 0) {
+		_run_snappy_app_dev_add_majmin(udev_s, nvpath, 195, 255);
+	}
+
+	sc_must_snprintf(nvpath, sizeof(nvpath), "/dev/nvidia%s", "-uvm");
+	if (stat(nvpath, &sbuf) == 0) {
+		_run_snappy_app_dev_add_majmin(udev_s, nvpath, 247, 0);
```

Comment:
Same suggestion about using something like: `const int nvidia_uvm_major = 247;`. It is pretty clear from the context, but I think I would still find it slightly easier to read this way.

Reply:
Done